src/kem/hqc/pqclean_hqc-128_clean/vector.c

```diff
@@ -0,0 +1,188 @@
+#include "parameters.h"
+#include "parsing.h"
+#include "randombytes.h"
+#include "vector.h"
+#include <stdint.h>
+#include <string.h>
+/**
+ * @file vector.c
+ * @brief Implementation of vectors sampling and some utilities for the HQC scheme
+ */
+
+static uint32_t m_val[75] = { 243079, 243093, 243106, 243120, 243134, 243148, 243161, 243175, 243189, 243203, 243216, 243230, 243244, 243258, 243272, 243285, 243299, 243313, 243327, 243340, 243354, 243368, 243382, 243396, 243409, 243423, 243437, 243451, 243465, 243478, 243492, 243506, 243520, 243534, 243547, 243561, 243575, 243589, 243603, 243616, 243630, 243644, 243658, 243672, 243686, 243699, 243713, 243727, 243741, 243755, 243769, 243782, 243796, 243810, 243824, 243838, 243852, 243865, 243879, 243893, 243907, 243921, 243935, 243949, 243962, 243976, 243990, 244004, 244018, 244032, 244046, 244059, 244073, 244087, 244101 };
+
+/**
+ * @brief Constant-time comparison of two integers v1 and v2
+ *
+ * Returns 1 if v1 is equal to v2 and 0 otherwise
+ * https://gist.github.com/sneves/10845247
+ *
+ * @param[in] v1
+ * @param[in] v2
+ */
+static inline uint32_t compare_u32(uint32_t v1, uint32_t v2) {
+    return 1 ^ ((uint32_t)((v1 - v2) | (v2 - v1)) >> 31);
+}
+
+static uint64_t single_bit_mask(uint32_t pos) {
+    uint64_t ret = 0;
+    uint64_t mask = 1;
+    uint64_t tmp;
+
+    for (size_t i = 0; i < 64; ++i) {
+        tmp = pos - i;
+        tmp = 0 - (1 - ((uint64_t)(tmp | (0 - tmp)) >> 63));
+        ret |= mask & tmp;
+        mask <<= 1;
+    }
+
+    return ret;
+}
+
+static inline uint32_t cond_sub(uint32_t r, uint32_t n) {
+    uint32_t mask;
+    r -= n;
+    mask = 0 - (r >> 31);
+    return r + (n & mask);
+}
+
+static inline uint32_t reduce(uint32_t a, size_t i) {
+    uint32_t q, n, r;
+    q = ((uint64_t) a * m_val[i]) >> 32;
```

Comment:
Just to double-check, since this is one of the constant-time patches: the index `i` isn't secret, right? Otherwise the lookup `m_val[i]` would be problematic.

Reply:
No, it's not. This routine is only called by `vect_set_random_fixed_weight`, where `i` is a loop variable ranging from 0 to a fixed public upper bound.

Reply:
ok thanks for the additional context